doc/explanation/package-management.md

```diff
@@ -0,0 +1,220 @@
+package lock.
+
+To do so, Dune uses
+[opam-0install-solver](https://github.com/ocaml-opam/opam-0install-solver),
+which is a variant of the 0install solver to find solutions for OPAM packages.
+
+Contrary to `opam`, the Dune solver always starts from a blank slate. It
+assumes nothing is installed and everything needs to be installed. This has the
+advantage that solving is simpler and previous solver solutions don't interfere
+with the current one. Thus given the same inputs it should always have the same
+outputs; no state is held between the solver runs.
+
+This can lead to more packages being installed (as OPAM won't by default
+install new versions of packages if the existing versions satisfy the
+constraints) but it avoids interference from already installed packages that
+lead to potentially different solutions.
+
+After solving is done, the solution gets written into the lock directory with
+all the data that is necessary to build and install the packages. From this
+point, the project contains all necessary information and does not need to
+access any package repositories.
+
+## Building
+
+When building, Dune will read the information from the lock directory and set
+up build rules to use the packages. The rules include
+
+* Fetch rules to download the sources as well as any additional sources like
+  patches and unpack them
+* Build rules to evaluate the build instructions from the build instructions
+  stored in the lock directory
+* Install rules to put the artifacts that were built into the appropriate,
+  Dune-managed folders
+
+Creating these processes as rules means that they will only be executed on
+demand, thus if the project has already downloaded the sources, it does not
+need to download them again. Likewise, if packages are installed, they stay
+installed.
+
+The results of the rules are stored in the `_build` directory of the project
+and managed automatically by Dune. Thus when cleaning the build directory, the
+installed packages are cleaned as well and will be reinstalled at the next
+build.
+
+When building the users project, the installed packages are added to the
+necessary search paths, thus user code can use the dependencies without any
+additional configuration.
+
+(compatibility)=
+## Packaging for Dune Compatibility
+
+Dune can build and install most packages as dependencies, even if they are not
+built with Dune themselves. Dune will execute the build instructions from the
+lock file, very similar to `opam`.
+
+However, there are some important restrictions that packages must adhere to to
+be able to work with Dune.
+
+The most important one is that the packages must not use absolute paths to
+refer to files. That means, they cannot read the path they are being built or
+installed into and expect this path to be stable. Dune builds packages in a
+sandbox location and after the build has finished it moves the files to the
+actual destination.
```

Comment:
Maybe a good place to explain the difference between Dune and Opam sandboxing models?

Reply:
The main difference is that we build in a different path than the one we install to. As far as I know the fact that we don't wrap things in `bwrap`/`sandbox-exec` are not design decisions, just something we haven't implemented yet. It is good idea and shouldn't be too difficult, especially with `opam` having made sure that packages generally work in its sandbox.

Reply:
I think its worth highlighting that we aren't bwrapping stuff – we can give a shoutout to Opam there as well.

Reply:
I don't really like to include things that can change easily within the scope of one PR as that will date the document and needs to be updated but I added a note on this.